server/channels/app/channel_bookmark.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+)
+
+func (a *App) GetChannelBookmarks(channelId string, since int64) ([]*model.ChannelBookmarkWithFileInfo, *model.AppError) {
+	bookmarks, err := a.Srv().Store().ChannelBookmark().GetBookmarksForChannelSince(channelId, since)
+	if err != nil {
+		return nil, model.NewAppError("GetChannelBookmarks", "app.channel.bookmark.get.app_error", nil, "", http.StatusNotFound).Wrap(err)
+	}
+
+	return bookmarks, nil
+}
+
+func (a *App) GetBookmark(bookmarkId string, includeDeleted bool) (*model.ChannelBookmarkWithFileInfo, *model.AppError) {
+	bookmark, err := a.Srv().Store().ChannelBookmark().Get(bookmarkId, includeDeleted)
+	if err != nil {
+		return nil, model.NewAppError("GetBookmark", "app.channel.bookmark.get.app_error", nil, "", http.StatusNotFound).Wrap(err)
+	}
+
+	return bookmark, nil
+}
+
+func (a *App) CreateChannelBookmark(c request.CTX, newBookmark *model.ChannelBookmark, connectionId string) (*model.ChannelBookmarkWithFileInfo, *model.AppError) {
+	newBookmark.OwnerId = c.Session().UserId
+	bookmark, err := a.Srv().Store().ChannelBookmark().Save(newBookmark, true)
+	if err != nil {
+		return nil, model.NewAppError("CreateChannelBookmark", "app.channel.bookmark.save.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	message := model.NewWebSocketEvent(model.WebsocketEventChannelBookmarkCreated, "", bookmark.ChannelId, "", nil, connectionId)
+	bookmarkJSON, jsonErr := json.Marshal(bookmark)
+	if jsonErr != nil {
+		return nil, model.NewAppError("CreateChannelBookmark", "api.marshal_error", nil, "", http.StatusInternalServerError).Wrap(jsonErr)
+	}
+	message.Add("bookmark", string(bookmarkJSON))
+	a.Publish(message)
+	return bookmark, nil
+}
+
+func (a *App) UpdateChannelBookmark(c request.CTX, updateBookmark *model.ChannelBookmarkWithFileInfo, connectionId string) (*model.UpdateChannelBookmarkResponse, *model.AppError) {
+	response := &model.UpdateChannelBookmarkResponse{}
+	if updateBookmark.OwnerId == c.Session().UserId {
+		isAnotherFile := updateBookmark.FileInfo != nil && updateBookmark.FileId != "" && updateBookmark.FileId != updateBookmark.FileInfo.Id
```

Comment:
What if the bookmark didn't have a file attached before? Wouldn't this `isAnotherFile` then be false and the new file wouldn't go through the necessary checks?

Reply:
The model is not valid in that scenario.

All bookmarks of type link must have a url, all bookmarks of type file must have a file

Reply:
But a bookmark of type link can also have a file, at least as far as I can tell

Reply:
done

Reply:
Looks like the Patch method can also change the bookmark type?